fdbserver/RestoreLoader.actor.cpp

```diff
@@ -550,38 +547,50 @@ ACTOR Future<Void> sendMutationsToApplier(VersionedMutationsMap* pkvOps, int bat
-			}
-			TraceEvent(SevDebug, "FastRestoreLoaderSendMutationToApplier")
-			    .detail("MessageIndex", msgIndex)
-			    .detail("RestoreAsset", asset.toString())
-			    .detail("Requests", requests.size());
-			wait(sendBatchRequests(&RestoreApplierInterface::sendMutationVector, *pApplierInterfaces, requests,
-			                       TaskPriority::RestoreLoaderSendMutations));
-			msgIndex++;
-			msgSize = 0;
-			requests.clear();
-			for (auto& applierID : applierIDs) {
-				applierMutationsBuffer[applierID] = MutationsVec();
-				applierVersionsBuffer[applierID] = LogMessageVersionVec();
-				applierMutationsSize[applierID] = 0.0;
+			// Batch mutations at multiple versions up to FASTRESTORE_LOADER_SEND_MUTATION_MSG_BYTES size
+			// to improve bandwidth from a loader to appliers
+			if (msgSize >= SERVER_KNOBS->FASTRESTORE_LOADER_SEND_MUTATION_MSG_BYTES) {
+				for (const UID& applierID : applierIDs) {
+					requests.emplace_back(applierID,
+					                      RestoreSendVersionedMutationsRequest(batchIndex, asset, msgIndex, isRangeFile,
+					                                                           applierMutationsBuffer[applierID],
+					                                                           applierVersionsBuffer[applierID]));
+				}
+				TraceEvent(SevDebug, "FastRestoreLoaderSendMutationToApplier")
+				    .detail("MessageIndex", msgIndex)
+				    .detail("RestoreAsset", asset.toString())
+				    .detail("Requests", requests.size());
+				wait(sendBatchRequests(&RestoreApplierInterface::sendMutationVector, *pApplierInterfaces, requests,
+				                       TaskPriority::RestoreLoaderSendMutations));
+				msgIndex++;
+				msgSize = 0;
+				requests.clear();
+				for (auto& applierID : applierIDs) {
+					applierMutationsBuffer[applierID] = MutationsVec();
+					applierVersionsBuffer[applierID] = LogMessageVersionVec();
+					applierMutationsSize[applierID] = 0.0;
+				}
 			}
-		}
+		} // Mutations at the same LogMessageVersion
 	} // all versions of mutations in the same file
 
+	// Send the remaining mutations in the applierMutationsBuffer
+	if (msgSize > 0) {
+		// TODO: Sanity check each asset has been received exactly once!
+		for (const UID& applierID : applierIDs) {
+			requests.emplace_back(applierID, RestoreSendVersionedMutationsRequest(
+			                                     batchIndex, asset, msgIndex, isRangeFile,
+			                                     applierMutationsBuffer[applierID], applierVersionsBuffer[applierID]));
+		}
+		TraceEvent(SevDebug, "FastRestoreLoaderSendMutationToApplier")
+		    .detail("MessageIndex", msgIndex)
+		    .detail("RestoreAsset", asset.toString())
+		    .detail("Requests", requests.size());
+		wait(sendBatchRequests(&RestoreApplierInterface::sendMutationVector, *pApplierInterfaces, requests,
+		                       TaskPriority::RestoreLoaderSendMutations));
+	}
```

Comment:
This part of code can be refactored into a function, since it was used twice. In addition, if the the buffer is empty, is there still a need to send the request?

Reply:
I changed this a lot by letting out of order sending. Using a function for this will not save code any more.